Problems with blockchain account and balances

## Problem Definition

1. Remove BTC account, works but is not reflected in the UI
2. Remove ETH account works but is not reflected in the UI and the loading bar is stuck
In console we get this:
```
vue.runtime.esm.js:619 [Vue warn]: Error in v-on handler (Promise/async): "Error: timeout of 30000ms exceeded"

found in

---> <ConfirmDialog> at src/components/dialogs/ConfirmDialog.vue
       <AccountBalances> at src/components/settings/AccountBalances.vue
         <VCard>
           <BlockchainBalances> at src/components/settings/BlockchainBalances.vue
             <AccountsBalances> at src/views/AccountsBalances.vue
               <VContent>
                 <VApp>
                   <App> at src/App.vue
                     <Root>
warn @ vue.runtime.esm.js:619
logError @ vue.runtime.esm.js:1884
globalHandleError @ vue.runtime.esm.js:1879
handleError @ vue.runtime.esm.js:1839
(anonymous) @ vue.runtime.esm.js:1856
Promise.catch (async)
invokeWithErrorHandling @ vue.runtime.esm.js:1856
invoker @ vue.runtime.esm.js:2179
invokeWithErrorHandling @ vue.runtime.esm.js:1854
Vue.$emit @ vue.runtime.esm.js:3888
emit @ vue-property-decorator.js:223
emitter @ vue-property-decorator.js:232
click @ ConfirmDialog.vue?9154:65
invokeWithErrorHandling @ vue.runtime.esm.js:1854
invoker @ vue.runtime.esm.js:2179
invokeWithErrorHandling @ vue.runtime.esm.js:1854
Vue.$emit @ vue.runtime.esm.js:3888
click @ VBtn.js:125
invokeWithErrorHandling @ vue.runtime.esm.js:1854
invoker @ vue.runtime.esm.js:2179
original._wrapper @ vue.runtime.esm.js:6917
vue.runtime.esm.js:1888 Error: timeout of 30000ms exceeded
    at createError (createError.js:16)
    at XMLHttpRequest.handleTimeout (xhr.js:89)
```

3. Adding Eth account:  `Error while adding account: Error: timeout of 30000ms exceeded
vue.runtime.esm.js:619 [Vue warn]: Error in v-on handler (Promise/async): "Error: timeout of 30000ms exceeded"`

4. After that and with the loading screen stuck removing an ETH token does not work


## Task

Solve them and perhaps write tests (if tests take time push them after the release as I can see various things needing fixing)

--- a/electron-app/src/components/settings/AccountAssetBalances.vue
+++ b/electron-app/src/components/settings/AccountAssetBalances.vue
@@ -5,6 +5,8 @@
         :items="tokens"
         :headers="headers"
         class="account-asset-balances__table"
+        sort-by="usdValue"
+        sort-desc
       >
         <template #item.asset="{ item }">
           <span class="account-asset-balances__balance__asset">

--- a/electron-app/src/components/settings/AccountBalances.vue
+++ b/electron-app/src/components/settings/AccountBalances.vue
@@ -31,7 +31,7 @@
     <v-data-table
       :headers="headers"
       :items="balances"
-      :loading="deleting || isLoading"
+      :loading="accountOperation || isLoading"
       loading-text="Please wait while Rotki queries the blockchain..."
       single-expand
       item-key="account"
@@ -54,7 +54,7 @@
       <template #item.actions="{ item }">
         <v-icon
           small
-          :disabled="deleting"
+          :disabled="accountOperation"
           @click="toDeleteAccount = item.account"
         >
           fa-trash
@@ -160,7 +160,7 @@ export default class AccountBalances extends Vue {
   hasTokens!: (account: string) => boolean;
 
   toDeleteAccount: string = '';
-  deleting = false;
+  pending = false;
 
   headers = [
     { text: 'Account', value: 'account' },
@@ -170,17 +170,25 @@ export default class AccountBalances extends Vue {
     { text: '', value: 'expand', align: 'end' }
   ];
 
+  get accountOperation(): boolean {
+    return (
+      this.isTaskRunning(TaskType.ADD_ACCOUNT) ||
+      this.isTaskRunning(TaskType.REMOVE_ACCOUNT) ||
+      this.pending
+    );
+  }
+
   async deleteAccount() {
     const address = this.toDeleteAccount;
     const blockchain = this.blockchain;
     this.toDeleteAccount = '';
-    this.deleting = true;
+    this.pending = true;
 
     await this.$store.dispatch('balances/removeAccount', {
       address,
       blockchain
     });
-    this.deleting = false;
+    this.pending = false;
   }
 
   refresh() {

--- a/electron-app/src/components/settings/AccountBalances.vue
+++ b/electron-app/src/components/settings/AccountBalances.vue
@@ -36,6 +36,8 @@
       single-expand
       item-key="account"
       :expanded.sync="expanded"
+      sort-by="usdValue"
+      sort-desc
     >
       <template #header.usdValue> {{ currency.ticker_symbol }} value </template>
       <template #item.account="{ item }">

--- a/electron-app/src/components/settings/AssetBalances.vue
+++ b/electron-app/src/components/settings/AssetBalances.vue
@@ -10,6 +10,8 @@
       :items="balances"
       :loading="isLoading"
       loading-text="Please wait while Rotki queries the blockchain..."
+      sort-by="usdValue"
+      sort-desc
     >
       <template #header.usdValue> {{ currency.ticker_symbol }} value </template>
       <template #item.asset="{ item }">

--- a/electron-app/src/components/settings/BlockchainBalances.vue
+++ b/electron-app/src/components/settings/BlockchainBalances.vue
@@ -1,4 +1,3 @@
-import { Severity } from "@/typing/types";
 <template>
   <v-row class="blockchain-balances">
     <v-col>
@@ -10,17 +9,20 @@ import { Severity } from "@/typing/types";
             class="blockchain-balances__chain"
             :items="items"
             label="Choose blockchain"
-            :disabled="loading"
+            :disabled="accountOperation"
           ></v-select>
           <v-text-field
             v-model="accountAddress"
             class="blockchain-balances__address"
             label="Account"
-            :disabled="loading"
+            :disabled="accountOperation"
           ></v-text-field>
 
           <div class="blockchain-balances--progress">
-            <v-progress-linear v-if="loading" indeterminate></v-progress-linear>
+            <v-progress-linear
+              v-if="accountOperation"
+              indeterminate
+            ></v-progress-linear>
           </div>
 
           <v-btn
@@ -28,7 +30,7 @@ import { Severity } from "@/typing/types";
             depressed
             color="primary"
             type="submit"
-            :disabled="!accountAddress || loading"
+            :disabled="!accountAddress || accountOperation"
             @click="addAccount()"
           >
             Add
@@ -69,12 +71,17 @@ import AccountBalances from '@/components/settings/AccountBalances.vue';
 import AssetBalances from '@/components/settings/AssetBalances.vue';
 import { notify } from '@/store/notifications/utils';
 import { Blockchain, Severity, SupportedBlockchains } from '@/typing/types';
+import { TaskType } from '@/model/task';
 
 const { mapGetters } = createNamespacedHelpers('balances');
+const { mapGetters: mapTaskGetters } = createNamespacedHelpers('tasks');
 
 @Component({
   components: { AccountBalances, AssetBalances, TokenTrack, MessageDialog },
-  computed: mapGetters(['ethAccounts', 'btcAccounts', 'totals'])
+  computed: {
+    ...mapGetters(['ethAccounts', 'btcAccounts', 'totals']),
+    ...mapTaskGetters(['isTaskRunning'])
+  }
 })
 export default class BlockchainBalances extends Vue {
   selected: Blockchain = 'ETH';
@@ -84,13 +91,21 @@ export default class BlockchainBalances extends Vue {
   ethAccounts!: AccountBalance[];
   btcAccounts!: AccountBalance[];
   totals!: AccountBalance[];
+  isTaskRunning!: (type: TaskType) => boolean;
+  pending: boolean = false;
 
-  loading: boolean = false;
+  get accountOperation(): boolean {
+    return (
+      this.isTaskRunning(TaskType.ADD_ACCOUNT) ||
+      this.isTaskRunning(TaskType.REMOVE_ACCOUNT) ||
+      this.pending
+    );
+  }
 
   async addAccount() {
     const address = this.accountAddress;
     const blockchain = this.selected;
-    this.loading = true;
+    this.pending = true;
     try {
       await this.$store.dispatch('balances/addAccount', {
         blockchain,
@@ -104,7 +119,7 @@ export default class BlockchainBalances extends Vue {
         Severity.ERROR
       );
     }
-    this.loading = false;
+    this.pending = false;
   }
 }
 </script>

--- a/electron-app/src/components/settings/FiatBalances.vue
+++ b/electron-app/src/components/settings/FiatBalances.vue
@@ -30,7 +30,12 @@
             {{ add ? 'Add Balance' : 'Modify Balance' }}
           </v-btn>
           <v-col cols="12">
-            <v-data-table :headers="headers" :items="fiatBalances">
+            <v-data-table
+              :headers="headers"
+              :items="fiatBalances"
+              sort-by="usdValue"
+              sort-desc
+            >
               <template #header.usdValue>
                 {{ currency.ticker_symbol }} value
               </template>

--- a/electron-app/src/model/task.ts
+++ b/electron-app/src/model/task.ts
@@ -30,6 +30,8 @@ export const createTask: <T extends TaskMeta>(
 });
 
 export enum TaskType {
+  ADD_ACCOUNT = 'add_account',
+  REMOVE_ACCOUNT = 'remove_account',
   TRADE_HISTORY = 'process_trade_history',
   QUERY_BLOCKCHAIN_BALANCES = 'query_blockchain_balances_async',
   QUERY_EXCHANGE_BALANCES = 'query_exchange_balances_async',

--- a/electron-app/src/services/rotkehlchen-api.ts
+++ b/electron-app/src/services/rotkehlchen-api.ts
@@ -816,11 +816,12 @@ export class RotkehlchenApi {
   removeBlockchainAccount(
     blockchain: string,
     account: string
-  ): Promise<BlockchainAccount> {
-    return new Promise<BlockchainAccount>((resolve, reject) => {
+  ): Promise<AsyncQuery> {
+    return new Promise<AsyncQuery>((resolve, reject) => {
       this.axios
-        .delete<ActionResult<BlockchainAccount>>(`/blockchains/${blockchain}`, {
+        .delete<ActionResult<AsyncQuery>>(`/blockchains/${blockchain}`, {
           data: {
+            async_query: true,
             accounts: [account]
           },
           validateStatus: function(status) {
@@ -844,12 +845,13 @@ export class RotkehlchenApi {
   addBlockchainAccount(
     blockchain: string,
     account: string
-  ): Promise<BlockchainAccount> {
-    return new Promise<BlockchainAccount>((resolve, reject) => {
+  ): Promise<AsyncQuery> {
+    return new Promise<AsyncQuery>((resolve, reject) => {
       this.axios
-        .put<ActionResult<BlockchainAccount>>(
+        .put<ActionResult<AsyncQuery>>(
           `/blockchains/${blockchain}`,
           {
+            async_query: true,
             accounts: [account]
           },
           {

--- a/electron-app/src/services/rotkehlchen-api.ts
+++ b/electron-app/src/services/rotkehlchen-api.ts
@@ -124,16 +124,16 @@ export class RotkehlchenApi {
 
   setPremiumCredentials(
     username: string,
-    api_key: string,
-    api_secret: string
+    apiKey: string,
+    apiSecret: string
   ): Promise<boolean> {
     return new Promise<boolean>((resolve, reject) => {
       this.axios
         .patch<ActionResult<boolean>>(
           `/users/${username}`,
           {
-            api_key,
-            api_secret
+            premium_api_key: apiKey,
+            premium_api_secret: apiSecret
           },
           { validateStatus: this.validate_status_patch_username }
         )

--- a/electron-app/src/services/task-manager.ts
+++ b/electron-app/src/services/task-manager.ts
@@ -6,7 +6,12 @@ import {
   convertBalances,
   convertEthBalances
 } from '@/utils/conversion';
-import { ExchangeMeta, TaskMeta, TaskType } from '@/model/task';
+import {
+  BlockchainMetadata,
+  ExchangeMeta,
+  TaskMeta,
+  TaskType
+} from '@/model/task';
 import { notify } from '@/store/notifications/utils';
 import { api } from '@/services/rotkehlchen-api';
 import {
@@ -55,6 +60,28 @@ export class TaskManager {
 
     store.commit('balances/updateEth', convertEthBalances(ETH));
     store.commit('balances/updateBtc', convertBalances(BTC));
+    store.commit('balances/updateTotals', convertBalances(totals));
+  }
+
+  onAccountOperation(
+    data: ActionResult<BlockchainBalances>,
+    meta: BlockchainMetadata
+  ) {
+    const { result, message } = data;
+    const { description, blockchain } = meta;
+
+    if (message) {
+      notify(`Operation failed due to ${message}`, description);
+      return;
+    }
+    const { per_account, totals } = result;
+    const { ETH, BTC } = per_account;
+
+    if (blockchain === 'ETH') {
+      store.commit('balances/updateEth', convertEthBalances(ETH));
+    } else {
+      store.commit('balances/updateBtc', convertBalances(BTC));
+    }
     store.commit('balances/updateTotals', convertBalances(totals));
   }
 
@@ -139,7 +166,9 @@ export class TaskManager {
   } = {
     [TaskType.QUERY_EXCHANGE_BALANCES]: TaskManager.onQueryExchangeBalances,
     [TaskType.QUERY_BLOCKCHAIN_BALANCES]: this.onQueryBlockchainBalances,
-    [TaskType.TRADE_HISTORY]: this.onTradeHistory
+    [TaskType.TRADE_HISTORY]: this.onTradeHistory,
+    [TaskType.ADD_ACCOUNT]: this.onAccountOperation,
+    [TaskType.REMOVE_ACCOUNT]: this.onAccountOperation
   };
 }
 

--- a/electron-app/src/store/balances/actions.ts
+++ b/electron-app/src/store/balances/actions.ts
@@ -12,7 +12,6 @@ import { Blockchain, Severity, UsdToFiatExchangeRates } from '@/typing/types';
 import { notify } from '@/store/notifications/utils';
 import { FiatBalance } from '@/model/blockchain-balances';
 import { bigNumberify } from '@/utils/bignumbers';
-import { convertBalances, convertEthBalances } from '@/utils/conversion';
 import { currencies } from '@/data/currencies';
 
 export const actions: ActionTree<BalanceState, RotkehlchenState> = {
@@ -173,33 +172,25 @@ export const actions: ActionTree<BalanceState, RotkehlchenState> = {
 
   async removeAccount({ commit }, payload: BlockchainAccountPayload) {
     const { address, blockchain } = payload;
-    const { per_account, totals } = await api.removeBlockchainAccount(
-      blockchain,
-      address
-    );
-    const { ETH, BTC } = per_account;
+    const { task_id } = await api.removeBlockchainAccount(blockchain, address);
 
-    if (blockchain === 'ETH') {
-      commit('updateEth', convertEthBalances(ETH));
-    } else {
-      commit('updateBtc', convertBalances(BTC));
-    }
-    commit('updateTotals', convertBalances(totals));
+    const task = createTask(task_id, TaskType.ADD_ACCOUNT, {
+      description: `Remove ${blockchain} account ${address}`,
+      blockchain
+    } as BlockchainMetadata);
+
+    commit('tasks/add', task, { root: true });
   },
 
   async addAccount({ commit }, payload: BlockchainAccountPayload) {
     const { address, blockchain } = payload;
-    const { per_account, totals } = await api.addBlockchainAccount(
-      blockchain,
-      address
-    );
-    const { ETH, BTC } = per_account;
-    if (blockchain === 'ETH') {
-      commit('updateEth', convertEthBalances(ETH));
-    } else {
-      commit('updateBtc', convertBalances(BTC));
-    }
-    commit('updateTotals', convertBalances(totals));
+    const { task_id } = await api.addBlockchainAccount(blockchain, address);
+    const task = createTask(task_id, TaskType.ADD_ACCOUNT, {
+      description: `Adding ${blockchain} account ${address}`,
+      blockchain
+    } as BlockchainMetadata);
+
+    commit('tasks/add', task, { root: true });
   }
 };
 

--- a/rotkehlchen/blockchain.py
+++ b/rotkehlchen/blockchain.py
@@ -622,8 +622,11 @@ class Blockchain(CacheableObject, LockableQueryObject):
             for account, balance in token_accounts.items():
                 token_total += balance
                 usd_value = balance * token_usd_price[token]
-                eth_balances[account][token] = balance
-                eth_balances[account]['usd_value'] = eth_balances[account]['usd_value'] + usd_value
+                if balance != ZERO:
+                    eth_balances[account][token] = balance
+                    eth_balances[account]['usd_value'] = (
+                        eth_balances[account]['usd_value'] + usd_value
+                    )
 
             self.totals[token] = {
                 'amount': token_total,

--- a/rotkehlchen/blockchain.py
+++ b/rotkehlchen/blockchain.py
@@ -311,7 +311,7 @@ class Blockchain(CacheableObject, LockableQueryObject):
         if append_or_remove == 'append':
             self.balances[A_BTC][account] = {'amount': balance, 'usd_value': usd_balance}
         elif append_or_remove == 'remove':
-            if account in self.balances[A_BTC][account]:
+            if account in self.balances[A_BTC]:
                 del self.balances[A_BTC][account]
         else:
             raise AssertionError('Programmer error: Should be append or remove')

--- a/rotkehlchen/tests/api/test_blockchain.py
+++ b/rotkehlchen/tests/api/test_blockchain.py
@@ -718,10 +718,8 @@ def test_remove_blockchain_accounts(
     removed_eth_accounts = [ethereum_accounts[0], ethereum_accounts[2]]
     eth_accounts_after_removal = [ethereum_accounts[1], ethereum_accounts[3]]
     all_eth_balances = ['1000000', '2000000', '3000000', '4000000']
-    # token_balances = {'RDN': ['0', '250000000', '450000000', '0']}
     token_balances = {'RDN': ['0', '0', '450000000', '0']}
     eth_balances_after_removal = ['2000000', '4000000']
-    # token_balances_after_removal = {'RDN': ['250000000', '0']}
     token_balances_after_removal = {'RDN': ['0', '0']}
     if query_balances_before_first_modification:
         # Also test by having balances queried before removing an account

--- a/rotkehlchen/tests/api/test_blockchain.py
+++ b/rotkehlchen/tests/api/test_blockchain.py
@@ -715,17 +715,21 @@ def test_remove_blockchain_accounts(
     rotki = rotkehlchen_api_server.rest_api.rotkehlchen
     rotki.blockchain.cache_ttl_secs = 0
 
+    removed_eth_accounts = [ethereum_accounts[0], ethereum_accounts[2]]
+    eth_accounts_after_removal = [ethereum_accounts[1], ethereum_accounts[3]]
+    all_eth_balances = ['1000000', '2000000', '3000000', '4000000']
+    # token_balances = {'RDN': ['0', '250000000', '450000000', '0']}
+    token_balances = {'RDN': ['0', '0', '450000000', '0']}
+    eth_balances_after_removal = ['2000000', '4000000']
+    # token_balances_after_removal = {'RDN': ['250000000', '0']}
+    token_balances_after_removal = {'RDN': ['0', '0']}
     if query_balances_before_first_modification:
         # Also test by having balances queried before removing an account
-        removed_eth_accounts = [ethereum_accounts[0], ethereum_accounts[2]]
-        all_eth_accounts = [ethereum_accounts[1], ethereum_accounts[3]]
-        eth_balances = ['3000000', '4000000']
-        token_balances = {'RDN': ['0', '250000000']}
         setup = setup_balances(
             rotki,
-            ethereum_accounts=all_eth_accounts,
+            ethereum_accounts=ethereum_accounts,
             btc_accounts=btc_accounts,
-            eth_balances=eth_balances,
+            eth_balances=all_eth_balances,
             token_balances=token_balances,
         )
         with setup.etherscan_patch, setup.bitcoin_patch:
@@ -733,17 +737,11 @@ def test_remove_blockchain_accounts(
                 rotkehlchen_api_server,
                 "blockchainbalancesresource",
             ))
-
-    # Accounts to remove. Assure they are not in order to test that it does not affect anything
-    removed_eth_accounts = [ethereum_accounts[0], ethereum_accounts[2]]
-    all_eth_accounts = [ethereum_accounts[1], ethereum_accounts[3]]
-    eth_balances = ['3000000', '4000000']
-    token_balances = {'RDN': ['0', '250000000']}
     setup = setup_balances(
         rotki,
-        ethereum_accounts=all_eth_accounts,
+        ethereum_accounts=ethereum_accounts,
         btc_accounts=btc_accounts,
-        eth_balances=eth_balances,
+        eth_balances=all_eth_balances,
         token_balances=token_balances,
     )
 
@@ -761,15 +759,15 @@ def test_remove_blockchain_accounts(
     assert_eth_balances_result(
         rotki=rotki,
         json_data=json_data,
-        eth_accounts=all_eth_accounts,
-        eth_balances=setup.eth_balances,
-        token_balances=setup.token_balances,
-        also_btc=False,  # All blockchain assets have not been queried yet
+        eth_accounts=eth_accounts_after_removal,
+        eth_balances=eth_balances_after_removal,
+        token_balances=token_balances_after_removal,
+        also_btc=query_balances_before_first_modification,
     )
     # Also make sure they are removed from the DB
     accounts = rotki.data.db.get_blockchain_accounts()
     assert len(accounts.eth) == 2
-    assert all(acc in accounts.eth for acc in all_eth_accounts)
+    assert all(acc in accounts.eth for acc in eth_accounts_after_removal)
     assert len(accounts.btc) == 2
     assert all(acc in accounts.btc for acc in btc_accounts)
 
@@ -785,21 +783,22 @@ def test_remove_blockchain_accounts(
     assert_eth_balances_result(
         rotki=rotki,
         json_data=json_data,
-        eth_accounts=all_eth_accounts,
-        eth_balances=setup.eth_balances,
-        token_balances=setup.token_balances,
+        eth_accounts=eth_accounts_after_removal,
+        eth_balances=eth_balances_after_removal,
+        token_balances=token_balances_after_removal,
         also_btc=True,
     )
 
     # Now we will try to remove a BTC account. Setup the mocking infrastructure again
-    all_btc_accounts = [UNIT_BTC_ADDRESS2]
+    all_btc_accounts = [UNIT_BTC_ADDRESS1, UNIT_BTC_ADDRESS2]
+    btc_accounts_after_removal = [UNIT_BTC_ADDRESS2]
     setup = setup_balances(
         rotki,
-        ethereum_accounts=all_eth_accounts,
+        ethereum_accounts=eth_accounts_after_removal,
         btc_accounts=all_btc_accounts,
-        eth_balances=eth_balances,
-        token_balances=token_balances,
-        btc_balances=['600000000'],
+        eth_balances=eth_balances_after_removal,
+        token_balances=token_balances_after_removal,
+        btc_balances=['3000000', '5000000'],
     )
     # remove the new BTC account
     with setup.bitcoin_patch:
@@ -814,16 +813,16 @@ def test_remove_blockchain_accounts(
     assert json_data['message'] == ''
     assert_btc_balances_result(
         json_data=json_data,
-        btc_accounts=all_btc_accounts,
-        btc_balances=setup.btc_balances,
+        btc_accounts=btc_accounts_after_removal,
+        btc_balances=['5000000'],
         also_eth=True,
     )
     # Also make sure it's removed from the DB
     accounts = rotki.data.db.get_blockchain_accounts()
     assert len(accounts.eth) == 2
-    assert all(acc in accounts.eth for acc in all_eth_accounts)
+    assert all(acc in accounts.eth for acc in eth_accounts_after_removal)
     assert len(accounts.btc) == 1
-    assert all(acc in accounts.btc for acc in all_btc_accounts)
+    assert all(acc in accounts.btc for acc in btc_accounts_after_removal)
 
     # Now try to query all balances to make sure the result is also stored
     with setup.etherscan_patch, setup.bitcoin_patch:
@@ -836,8 +835,8 @@ def test_remove_blockchain_accounts(
     assert json_data['message'] == ''
     assert_btc_balances_result(
         json_data=json_data,
-        btc_accounts=all_btc_accounts,
-        btc_balances=setup.btc_balances,
+        btc_accounts=btc_accounts_after_removal,
+        btc_balances=['5000000'],
         also_eth=True,
     )
 
@@ -860,14 +859,16 @@ def test_remove_blockchain_accounts_async(
 
     # Test by having balances queried before removing an account
     removed_eth_accounts = [ethereum_accounts[0], ethereum_accounts[2]]
-    all_eth_accounts = [ethereum_accounts[1], ethereum_accounts[3]]
-    eth_balances = ['3000000', '4000000']
-    token_balances = {'RDN': ['0', '250000000']}
+    eth_accounts_after_removal = [ethereum_accounts[1], ethereum_accounts[3]]
+    all_eth_balances = ['1000000', '2000000', '3000000', '4000000']
+    token_balances = {'RDN': ['0', '250000000', '450000000', '0']}
+    eth_balances_after_removal = ['2000000', '4000000']
+    token_balances_after_removal = {'RDN': ['250000000', '0']}
     setup = setup_balances(
         rotki,
-        ethereum_accounts=all_eth_accounts,
+        ethereum_accounts=ethereum_accounts,
         btc_accounts=btc_accounts,
-        eth_balances=eth_balances,
+        eth_balances=all_eth_balances,
         token_balances=token_balances,
     )
     with setup.etherscan_patch, setup.bitcoin_patch:
@@ -875,17 +876,11 @@ def test_remove_blockchain_accounts_async(
             rotkehlchen_api_server,
             "blockchainbalancesresource",
         ))
-
-    # Accounts to remove. Assure they are not in order to test that it does not affect anything
-    removed_eth_accounts = [ethereum_accounts[0], ethereum_accounts[2]]
-    all_eth_accounts = [ethereum_accounts[1], ethereum_accounts[3]]
-    eth_balances = ['3000000', '4000000']
-    token_balances = {'RDN': ['0', '250000000']}
     setup = setup_balances(
         rotki,
-        ethereum_accounts=all_eth_accounts,
+        ethereum_accounts=ethereum_accounts,
         btc_accounts=btc_accounts,
-        eth_balances=eth_balances,
+        eth_balances=all_eth_balances,
         token_balances=token_balances,
     )
 
@@ -901,15 +896,15 @@ def test_remove_blockchain_accounts_async(
     assert_eth_balances_result(
         rotki=rotki,
         json_data=outcome,
-        eth_accounts=all_eth_accounts,
-        eth_balances=setup.eth_balances,
-        token_balances=setup.token_balances,
-        also_btc=False,  # All blockchain assets have not been queried yet
+        eth_accounts=eth_accounts_after_removal,
+        eth_balances=eth_balances_after_removal,
+        token_balances=token_balances_after_removal,
+        also_btc=True,  # We queried all balances at the start
     )
     # Also make sure they are removed from the DB
     accounts = rotki.data.db.get_blockchain_accounts()
     assert len(accounts.eth) == 2
-    assert all(acc in accounts.eth for acc in all_eth_accounts)
+    assert all(acc in accounts.eth for acc in eth_accounts_after_removal)
     assert len(accounts.btc) == 2
     assert all(acc in accounts.btc for acc in btc_accounts)
 
@@ -925,21 +920,22 @@ def test_remove_blockchain_accounts_async(
     assert_eth_balances_result(
         rotki=rotki,
         json_data=json_data,
-        eth_accounts=all_eth_accounts,
-        eth_balances=setup.eth_balances,
-        token_balances=setup.token_balances,
+        eth_accounts=eth_accounts_after_removal,
+        eth_balances=eth_balances_after_removal,
+        token_balances=token_balances_after_removal,
         also_btc=True,
     )
 
     # Now we will try to remove a BTC account. Setup the mocking infrastructure again
-    all_btc_accounts = [UNIT_BTC_ADDRESS2]
+    all_btc_accounts = [UNIT_BTC_ADDRESS1, UNIT_BTC_ADDRESS2]
+    btc_accounts_after_removal = [UNIT_BTC_ADDRESS2]
     setup = setup_balances(
         rotki,
-        ethereum_accounts=all_eth_accounts,
+        ethereum_accounts=eth_accounts_after_removal,
         btc_accounts=all_btc_accounts,
-        eth_balances=eth_balances,
-        token_balances=token_balances,
-        btc_balances=['600000000'],
+        eth_balances=eth_balances_after_removal,
+        token_balances=token_balances_after_removal,
+        btc_balances=['3000000', '5000000'],
     )
     # remove the new BTC account
     with setup.bitcoin_patch:
@@ -952,16 +948,16 @@ def test_remove_blockchain_accounts_async(
         outcome = wait_for_async_task(rotkehlchen_api_server, task_id)
     assert_btc_balances_result(
         json_data=outcome,
-        btc_accounts=all_btc_accounts,
-        btc_balances=setup.btc_balances,
+        btc_accounts=btc_accounts_after_removal,
+        btc_balances=['5000000'],
         also_eth=True,
     )
     # Also make sure it's removed from the DB
     accounts = rotki.data.db.get_blockchain_accounts()
     assert len(accounts.eth) == 2
-    assert all(acc in accounts.eth for acc in all_eth_accounts)
+    assert all(acc in accounts.eth for acc in eth_accounts_after_removal)
     assert len(accounts.btc) == 1
-    assert all(acc in accounts.btc for acc in all_btc_accounts)
+    assert all(acc in accounts.btc for acc in btc_accounts_after_removal)
 
     # Now try to query all balances to make sure the result is also stored
     with setup.etherscan_patch, setup.bitcoin_patch:
@@ -974,7 +970,7 @@ def test_remove_blockchain_accounts_async(
     assert json_data['message'] == ''
     assert_btc_balances_result(
         json_data=json_data,
-        btc_accounts=all_btc_accounts,
-        btc_balances=setup.btc_balances,
+        btc_accounts=btc_accounts_after_removal,
+        btc_balances=['5000000'],
         also_eth=True,
     )

--- a/rotkehlchen/tests/utils/blockchain.py
+++ b/rotkehlchen/tests/utils/blockchain.py
@@ -246,10 +246,12 @@ def assert_btc_balances_result(
     msg = 'given balances and accounts should have same length'
     assert len(btc_accounts) == len(btc_balances), msg
     for idx, account in enumerate(btc_accounts):
-        assert FVal(per_account[account]['amount']) == satoshis_to_btc(
-            FVal(btc_balances[idx]),
-        )
-        assert FVal(per_account[account]['usd_value']) > ZERO
+        balance = satoshis_to_btc(FVal(btc_balances[idx]))
+        assert FVal(per_account[account]['amount']) == balance
+        if balance == ZERO:
+            assert FVal(per_account[account]['usd_value']) == ZERO
+        else:
+            assert FVal(per_account[account]['usd_value']) > ZERO
 
     totals = result['totals']
     if also_eth:
@@ -259,7 +261,10 @@ def assert_btc_balances_result(
 
     expected_btc_total = sum(satoshis_to_btc(FVal(balance)) for balance in btc_balances)
     assert FVal(totals['BTC']['amount']) == expected_btc_total
-    assert FVal(totals['BTC']['usd_value']) > ZERO
+    if expected_btc_total == ZERO:
+        assert FVal(totals['BTC']['usd_value']) == ZERO
+    else:
+        assert FVal(totals['BTC']['usd_value']) > ZERO
 
 
 def assert_eth_balances_result(
@@ -285,8 +290,12 @@ def assert_eth_balances_result(
         per_account = per_account['ETH']
         assert len(per_account) == len(eth_accounts)
         for idx, account in enumerate(eth_accounts):
-            assert FVal(per_account[account]['ETH']) == from_wei(FVal(eth_balances[idx]))
-            assert FVal(per_account[account]['usd_value']) > ZERO
+            balance = from_wei(FVal(eth_balances[idx]))
+            assert FVal(per_account[account]['ETH']) == balance
+            if balance == ZERO:
+                assert FVal(per_account[account]['usd_value']) == ZERO
+            else:
+                assert FVal(per_account[account]['usd_value']) > ZERO
             for symbol, balances in token_balances.items():
                 token_balance = FVal(balances[idx])
                 if token_balance != ZERO:
@@ -306,7 +315,11 @@ def assert_eth_balances_result(
 
     expected_total_eth = sum(from_wei(FVal(balance)) for balance in eth_balances)
     assert FVal(totals['ETH']['amount']) == expected_total_eth
-    assert FVal(totals['ETH']['usd_value']) > ZERO
+    if expected_total_eth == ZERO:
+        FVal(totals['ETH']['usd_value']) == ZERO
+    else:
+        assert FVal(totals['ETH']['usd_value']) > ZERO
+
     for symbol, balances in token_balances.items():
         if symbol not in rotki.blockchain.owned_eth_tokens:
             # If the token got removed from the owned tokens in the test make sure

--- a/rotkehlchen/tests/utils/blockchain.py
+++ b/rotkehlchen/tests/utils/blockchain.py
@@ -298,7 +298,10 @@ def assert_eth_balances_result(
                 assert FVal(per_account[account]['usd_value']) > ZERO
             for symbol, balances in token_balances.items():
                 token_balance = FVal(balances[idx])
-                if token_balance != ZERO:
+                if token_balance == ZERO:
+                    msg = f'{account} should have no entry for {symbol}'
+                    assert symbol not in per_account[account], msg
+                else:
                     assert FVal(per_account[account][symbol]) == from_wei(token_balance)
 
     if totals_only:

--- a/rotkehlchen/tests/utils/rotkehlchen.py
+++ b/rotkehlchen/tests/utils/rotkehlchen.py
@@ -79,7 +79,7 @@ def setup_balances(
             'The btc balances should be a list with each '
             'element representing balance of an account'
         )
-        assert len(eth_balances) == len(ethereum_accounts)
+        assert len(btc_balances) == len(btc_accounts)
     else:
         # Default test values
         if len(btc_accounts) != 0:

